### ISSUE TITLE
Valid version said to be invalid 

### ISSUE THREAD

**Issue by potatoqualitee** [potatoqualitee/psmodulecache#37]
@LaurentDardenne do you have time this week to check this out? Usually I'd be happy to but I'm trying to get a new dbatools library out. A valid version is being said to be invalid

```
The syntax of the version '2022.10.25.1' is invalid for 'dbatools-library:2022.10.25.1'.
The syntax of the version '2022.10.25.1' is invalid for 'dbatools-core-library:2022.10.25.1'
```

Also, I cannot get `dbatools-library::` to work, even with single quotes around it. Any suggestions?

**Comment by LaurentDardenne** [comment]
A priori the string '2022.10.25.1' is not a valid [semver](https://semver.org/) : 
```
2. A normal version number MUST take the form X.Y.Z where X, Y, and Z are non-negative integers, 
 and MUST NOT contain leading zeroes. X is the major version, Y is the minor version, and Z is the patch version.
Each element MUST increase numerically. For instance: 1.9.0 -> 1.10.0 -> 1.11.0.
```
With ''2022.10.25.1' the regex return $false .

I made the choice to use semver instead of [version], we can combine the two if we stay on 3 digits in the number.
The [Version] class does not know how to manage the pre-release part, but 4 digit:
```powershell
[version]'2022.10.25.1'
#ok
```
It's a format mixing puzzle and PSget seems to straddle both formats :-/
This test case does not exist.

Fixing this requires handling both formats in the [version test](https://github.com/potatoqualitee/psmodulecache/blob/main/PSModuleCache.psm1#L61).

It is necessary to manage the class [Version] OR the regex semver. 
_Either it's semver with or without prerelease, or a version without prerelease_
I'm looking at the test data this weekend.

**Comment by LaurentDardenne** [comment]
I adapted the test, as indicated above the versioning of the publication numbers can be done on both formats.
There are therefore some collisions to manage, a valid number for one of the formats may be invalid for the other.
```powershell
[String[]]$global:ValidSemanticVersions = @(
   #'1.2' invalid Semver but valid CLR version
   #'1' invalid for CLR version AND Semver
   '0.0.0' #valid CLR version
   '0.0.4', #valid CLR version
   '1.2.3', #valid CLR version
   '10.20.30', #valid CLR version
   '1.0.0', #valid CLR version
   '2.0.0', #valid CLR version
   '1.1.7', #valid CLR version
   ... #invalid CLR version
)

[String[]]$global:InvalidSemanticVersions = @(
   '0.0.0.0' #valid CLR version
   '0.0.-0' #valid CLR version !!
   '01.1.1', #valid CLR version
   '1.01.1', #valid CLR version
   '1.1.01', #valid CLR version
   '1.2', #valid CLR version
   '2022.1.2.3', #valid CLR version
   '1.0'  #valid CLR version
   '1.2', #valid CLR version
   ... #invalid semver
)

[String[]]$global:ValidClrVersions = @(
   #'1','11', '1.', '-1' are invalid for [version]'string'

   #'1.0' is valid but returns digits initialized to -1
   #  Major  Minor  Build  Revision
   #  -----  -----  -----  --------
   #  1      0      -1     -1
   #note : The value of Version properties that have not been explicitly assigned a value is undefined (-1).

   #[version]'1.0.-1' throw an exception
   #[version]'0.0.-0'  -> OK !!!
   '0.0.0.0'
   '1.2',
   '1.2.3',
   '2022.1.2.3' #It is a invalid Semver but a valid clr version.
)
```
There are also cases of the 'ValidSemanticVersionConstraints' test set to be excluded in the Test-Version and Test-PrereleaseVersion functions.
We will not find these syntaxes in a nuget repository, but in the value of the parameter it remains valid.
Finally for the syntax of a version number with prerelease, I don't know which ones are accepted by powershellget. That said, we only manage existing version numbers, therefore authorized on PSGallery.
I will try to submit a PR this WE...

**Comment by potatoqualitee** [comment]
Thank you so much for the accommodation!

**Comment by LaurentDardenne** [comment]
The problem seems to be [fixed](https://github.com/LaurentDardenne/psmodulecache/actions/runs/3356637250):
![image](https://user-images.githubusercontent.com/4894452/198898322-e87344f5-e8ee-4033-9d2d-282475926f9f.png)


However I still have to perform a check on the _'ConvertTo-Version_' function.

**Comment by LaurentDardenne** [comment]
I found [this](https://learn.microsoft.com/en-us/powershell/scripting/gallery/concepts/module-prerelease-support?view=powershell-5.1) last night and it confirms my doubts about whether or not to follow [PowershellGet version management](https://github.com/PowerShell/PowerShellGetv2/blob/master/src/PowerShellGet/private/functions/ValidateAndGet-VersionPrereleaseStrings.ps1).
As I use a regex for semver v2, the minimal risk is to allow valid prerelease version numbers for PSModuleCache but which cannot exist in PSGallery.
As a user, you enter an existing version number, except typing error.

I wonder if I'm not going to simplify this version management by reusing the powershell function mentioned above...

**Comment by potatoqualitee** [comment]
ohh, that is interesting! reusing their own command seems ideal

**Comment by potatoqualitee** [comment]
check it out, @LaurentDardenne ! We can now manage our caches

![image](https://user-images.githubusercontent.com/8278033/199947358-db91abb4-80f5-495d-9cd2-7fead485c480.png)

**Comment by LaurentDardenne** [comment]
Yes I saw but in an automation it does not help me, but we can already check the generation of the key name, which can help PSModuleCache users and others :-)

For this version problem which is to manage a specific [Version] + prerelease information I wanted to check if an external repository can contain modules with a different version format.
That is to say delivered without using the Publish-Module cmdlet which controls the version format authorized on PSGallery.
Then if we try to deliver a module with a 1.0 version, this cmdlet changes the version to 1.0.0, but there are modules on PSgallery with a 1.0 version delivered with a previous version of PSGet (I suppose).
It is unlikely, but not impossible, to require such a version to be cached. It is also to be tested.

But being on a complicated professional project, I don't have the desire to code the Weekend at the moment. I would do that on the fly.
If the management of version numbers '1.2.3.4' matters, we can however deliver what I had corrected, but just a patch in DIY mode ( pas sûr de la traduction :-) ).

**Comment by LaurentDardenne** [comment]
I just tested, so we can have any version on a module repository but only those respecting the PowershellGet versioning name can be managed :

![image](https://user-images.githubusercontent.com/4894452/200114887-43d9939c-bf99-4530-a807-c93fafe51f3c.png)

Modules in version 1.0 can be saved.
This seemed obvious, but we can now specify this behavior in the documentation.
I will adapt the code using the function mentioned above.

**Comment by LaurentDardenne** [comment]
I plan to fix this problem in the next few days.
To note :
The following version numbers are valid but do not follow the standard defined for PowershellGet.
Their use does not trigger a syntax error but returns 'NoMatchFoundForCriteria' :
```
'1.2-',
'1.2.3-',
'1.2.3--test',
'1.2.3.4-',
'1.2.3.4--',
'1.2.3- Beta1'
```

This syntax causes a bug when calling -split on a constrained variable:
```Powershell
[ValidateNotNullOrEmpty()][string]$Version='-1.2.3-'
$Version,$Prerelease = $Version -split '-',2
#Error: The variable cannot be validated because the value is not a valid value for the Version variable.
```
I leave [the function](https://github.com/PowerShell/PowerShellGetv2/blob/master/src/PowerShellGet/private/functions/ValidateAndGet-VersionPrereleaseStrings.ps1) code as is, it will remain consistent with PowershellGet.

**Comment by potatoqualitee** [comment]
this is fixed in the version i just merged?

**Comment by LaurentDardenne** [comment]
Yes it is fixed.
As stated I haven't fixed the bugs in the code of the 'ValidateAndGet-VersionPrereleaseStrings' function. And I think most of the time it doesn't bother anyone.

**Comment by LaurentDardenne** [comment]
In addition.
When searching for a dependent module, it seems that version '2.0.0' (key '_RequiredVersion_') is not considered the same as '2.0'...

```powershell
    RequiredModules = @( @{ ModuleName = 'modulerequirelicenseacceptance'; RequiredVersion = '2.0.0'; Guid = '8dba4e13-19fd-41b5-8840-246e2f2b30c7' } )
```

```powershell
# Version number of this module.
ModuleVersion = '2.0'

GUID = '8dba4e13-19fd-41b5-8840-246e2f2b30c7'

# Author of this module
Author = 'farehar'
```
![image](https://github.com/potatoqualitee/psmodulecache/assets/4894452/79a122f3-a2bf-47ef-9f87-e103a4da0dce)

C'est bien dommage !
